File: interview/agent-node/agent.ts
/**
 * LiveKit Voice Agent for Trivia Game (Node.js)
 * This agent hosts a trivia game where users can answer questions via voice.
 */

import {
  type JobContext,
  WorkerOptions,
  cli,
  defineAgent,
  voice,
  llm,
} from '@livekit/agents';
import * as openai from '@livekit/agents-plugin-openai';
import { BackgroundVoiceCancellation } from '@livekit/noise-cancellation-node';
import { fileURLToPath } from 'node:url';
import dotenv from 'dotenv';
import { TRIVIA_QUESTIONS, shuffleArray, type TriviaQuestion } from './trivia-data.js';
import { z } from 'zod';

dotenv.config();

// Backend API URL for score updates
const BACKEND_URL = process.env.BACKEND_URL || 'http://localhost:3001';

/**
 * Maintains the state of the trivia game
 */
class TriviaState {
  score: number = 0;
  questionsAsked: number = 0;
  currentQuestion: TriviaQuestion | null = null;
  availableQuestions: TriviaQuestion[];
  participantIdentity: string | null = null;
  gameActive: boolean = true;

  constructor() {
    this.availableQuestions = shuffleArray([...TRIVIA_QUESTIONS]);
  }

  getNextQuestion(): TriviaQuestion {
    if (this.availableQuestions.length === 0) {
      // Reshuffle if we run out
      this.availableQuestions = shuffleArray([...TRIVIA_QUESTIONS]);
    }

    this.currentQuestion = this.availableQuestions.pop()!;
    this.questionsAsked++;
    return this.currentQuestion;
  }

  checkAnswer(userAnswer: string): boolean {
    if (!this.currentQuestion) {
      return false;
    }

    const userAnswerLower = userAnswer.toLowerCase().trim();
    const correctAnswerLower = this.currentQuestion.answer.toLowerCase();

    // Check exact match
    if (userAnswerLower === correctAnswerLower) {
      return true;
    }

    // Check alternatives
    for (const alt of this.currentQuestion.alternatives) {
      if (userAnswerLower === alt.toLowerCase()) {
        return true;
      }
    }

    // Check if answer is contained in user's response (more lenient)
    if (userAnswerLower.includes(correctAnswerLower)) {
      return true;
    }

    return false;
  }

  async updateBackendScore(): Promise<void> {
    try {
      const payload = {
        participant_id: this.participantIdentity || 'guest',
        score: this.score,
        questions_asked: this.questionsAsked,
      };

      const response = await fetch(`${BACKEND_URL}/api/trivia/score`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
      });

      if (response.ok) {
        console.log('Score updated:', payload);
      } else {
        console.warn('Failed to update score:', response.status);
      }
    } catch (error) {
      console.error('Error updating backend score:', error);
    }
  }
}

/**
 * Voice agent that hosts the trivia game
 */
class TriviaHostAgent extends voice.Agent<TriviaState> {
  constructor(triviaState: TriviaState) {
    super({
      instructions: `You are an enthusiastic and friendly trivia game host.
Your role is to:
1. Welcome players warmly and explain the game
2. Ask trivia questions one at a time
3. Listen to the player's answer
4. Tell them if they're correct or incorrect
5. Keep track of their score
6. Be encouraging and fun!

When asking questions, speak clearly and wait for the user's response.
After they answer, immediately tell them if they're right or wrong.
Keep responses concise and engaging. Don't use asterisks or emojis in speech.
After giving feedback, ask if they want another question.`,
      tools: {
        ask_trivia_question: llm.tool({
          description: 'Ask the next trivia question to the user',
          execute: async (_, { ctx }: llm.ToolOptions<TriviaState>) => {
            const questionData = ctx.userData.getNextQuestion();
            const questionText = `Here's question number ${ctx.userData.questionsAsked}. Category: ${questionData.category}. ${questionData.question}`;

            console.log(`Asking question: ${questionData.question}`);
            console.log(`Correct answer: ${questionData.answer}`);

            return `Ask the user: '${questionText}'. Then wait for their answer.`;
          },
        }),

        check_user_answer: llm.tool({
          description:
            "Check if the user's answer is correct and update their score. Call this when the user provides an answer to a trivia question.",
          parameters: z.object({
            user_answer: z.string().describe('The answer provided by the user'),
          }),
          execute: async (
            { user_answer }: { user_answer: string },
            { ctx }: llm.ToolOptions<TriviaState>,
          ) => {
            const isCorrect = ctx.userData.checkAnswer(user_answer);

            let feedback: string;
            if (isCorrect) {
              ctx.userData.score++;
              feedback = `That's correct! The answer is ${ctx.userData.currentQuestion?.answer}. Great job! Your score is now ${ctx.userData.score} out of ${ctx.userData.questionsAsked}.`;
            } else {
              feedback = `Sorry, that's not quite right. The correct answer is ${ctx.userData.currentQuestion?.answer}. Your score is ${ctx.userData.score} out of ${ctx.userData.questionsAsked}.`;
            }

            console.log(
              `User answered: ${user_answer} - ${isCorrect ? 'Correct' : 'Incorrect'}`,
            );
            console.log(`Current score: ${ctx.userData.score}/${ctx.userData.questionsAsked}`);

            // Update backend
            await ctx.userData.updateBackendScore();

            return feedback;
          },
        }),

        get_current_score: llm.tool({
          description: "Get the user's current score",
          execute: async (_, { ctx }: llm.ToolOptions<TriviaState>) => {
            return `Your current score is ${ctx.userData.score} out of ${ctx.userData.questionsAsked} questions.`;
          },
        }),

        end_game: llm.tool({
          description: 'End the trivia game and show final score',
          execute: async (_, { ctx }: llm.ToolOptions<TriviaState>) => {
            ctx.userData.gameActive = false;
            const percentage =
              ctx.userData.questionsAsked > 0
                ? (ctx.userData.score / ctx.userData.questionsAsked) * 100
                : 0;

            const finalMessage = `Thanks for playing! Your final score is ${ctx.userData.score} out of ${ctx.userData.questionsAsked}. That's ${percentage.toFixed(1)} percent correct. Great job!`;

            await ctx.userData.updateBackendScore();

            return finalMessage;
          },
        }),
      },
    });
  }

  async onEnter(): Promise<void> {
    // Welcome message
    await this.session.say(
      `Welcome to Trivia Time! I'm your host and I'm excited to test your knowledge. 
I'll ask you questions from different categories like Science, History, Geography, Movies, and Sports. 
Just answer with what you think is correct. Ready to start? Let me ask you the first question!`,
      { allowInterruptions: false },
    );

    // Ask first question automatically
    await new Promise((resolve) => setTimeout(resolve, 500));
    const questionData = this.session.userData.getNextQuestion();
    const questionText = `Question number ${this.session.userData.questionsAsked}. Category: ${questionData.category}. ${questionData.question}`;

    console.log(`First question: ${questionData.question}`);
    console.log(`Correct answer: ${questionData.answer}`);

    await this.session.say(questionText);
  }
}

/**
 * Define the agent
 */
export default defineAgent({
  entry: async (ctx: JobContext) => {
    // Create trivia state for this session
    const triviaState = new TriviaState();

    // Connect to room
    await ctx.connect();

    // Get participant identity if available
    const participants = Array.from(ctx.room.remoteParticipants.values());
    if (participants.length > 0) {
      triviaState.participantIdentity = participants[0].identity;
      console.log(`Participant joined: ${participants[0].identity}`);
    }

    // Create agent session with OpenAI Realtime
    const session = new voice.AgentSession({
      llm: new openai.realtime.RealtimeModel({
        voice: 'alloy', // Friendly voice for trivia host
        temperature: 0.8,
      }),
      userData: triviaState,
    });

    // Start the session
    await session.start({
      agent: new TriviaHostAgent(triviaState),
      room: ctx.room,
      inputOptions: {
        noiseCancellation: BackgroundVoiceCancellation(),
      },
    });

    console.log('Trivia agent session started');
  },
});

// Run the agent
cli.runApp(new WorkerOptions({ agent: fileURLToPath(import.meta.url) }));

File: interview/agent/agent.py
"""
LiveKit Voice Agent for Trivia Game
This agent hosts a trivia game where users can answer questions via voice.
"""
import logging
import random
import asyncio
import aiohttp
from pathlib import Path
from dotenv import load_dotenv
from livekit import agents, rtc
from livekit.agents import AgentServer, AgentSession, Agent, room_io
from livekit.agents.llm import function_tool
from livekit.plugins import openai, noise_cancellation
from trivia_data import TRIVIA_QUESTIONS

load_dotenv(Path(__file__).parent / ".env")

logger = logging.getLogger("trivia-agent")
logger.setLevel(logging.INFO)

# Backend API URL for score updates
BACKEND_URL = "http://localhost:3001"

class TriviaState:
    """Maintains the state of the trivia game"""
    def __init__(self):
        self.score = 0
        self.questions_asked = 0
        self.current_question = None
        self.available_questions = TRIVIA_QUESTIONS.copy()
        random.shuffle(self.available_questions)
        self.participant_identity = None
        self.game_active = True
    
    def get_next_question(self):
        """Get the next random question"""
        if not self.available_questions:
            # Reshuffle if we run out
            self.available_questions = TRIVIA_QUESTIONS.copy()
            random.shuffle(self.available_questions)
        
        self.current_question = self.available_questions.pop()
        self.questions_asked += 1
        return self.current_question
    
    def check_answer(self, user_answer: str) -> bool:
        """Check if the user's answer is correct"""
        if not self.current_question:
            return False
        
        user_answer_lower = user_answer.lower().strip()
        correct_answer_lower = self.current_question["answer"].lower()
        
        # Check exact match or alternatives
        if user_answer_lower == correct_answer_lower:
            return True
        
        # Check alternatives
        for alt in self.current_question.get("alternatives", []):
            if user_answer_lower == alt.lower():
                return True
        
        # Check if answer is contained in user's response (more lenient)
        if correct_answer_lower in user_answer_lower:
            return True
            
        return False


class TriviaHostAgent(Agent):
    """Voice agent that hosts the trivia game"""
    
    def __init__(self, trivia_state: TriviaState) -> None:
        super().__init__(
            instructions="""You are an enthusiastic and friendly trivia game host.
            Your role is to:
            1. Welcome players warmly and explain the game
            2. Ask trivia questions one at a time
            3. Listen to the player's answer
            4. Tell them if they're correct or incorrect
            5. Keep track of their score
            6. Be encouraging and fun!
            
            When asking questions, speak clearly and wait for the user's response.
            After they answer, immediately tell them if they're right or wrong.
            Keep responses concise and engaging. Don't use asterisks or emojis in speech.
            After giving feedback, ask if they want another question.
            """,
        )
        self.trivia_state = trivia_state
    
    async def update_backend_score(self):
        """Send score update to the backend API"""
        try:
            async with aiohttp.ClientSession() as session:
                payload = {
                    "participant_id": self.trivia_state.participant_identity or "guest",
                    "score": self.trivia_state.score,
                    "questions_asked": self.trivia_state.questions_asked
                }
                async with session.post(
                    f"{BACKEND_URL}/api/trivia/score",
                    json=payload,
                    timeout=aiohttp.ClientTimeout(total=5)
                ) as response:
                    if response.status == 200:
                        logger.info(f"Score updated: {payload}")
                    else:
                        logger.warning(f"Failed to update score: {response.status}")
        except Exception as e:
            logger.error(f"Error updating backend score: {e}")
    
    @function_tool
    async def ask_trivia_question(self):
        """Ask the next trivia question to the user"""
        question_data = self.trivia_state.get_next_question()
        question_text = f"Here's question number {self.trivia_state.questions_asked}. Category: {question_data['category']}. {question_data['question']}"
        
        logger.info(f"Asking question: {question_data['question']}")
        logger.info(f"Correct answer: {question_data['answer']}")
        
        return None, f"Ask the user: '{question_text}'. Then wait for their answer."
    
    @function_tool
    async def check_user_answer(self, user_answer: str):
        """
        Check if the user's answer is correct and update their score.
        
        Args:
            user_answer: The answer provided by the user
        """
        is_correct = self.trivia_state.check_answer(user_answer)
        
        if is_correct:
            self.trivia_state.score += 1
            feedback = f"That's correct! The answer is {self.trivia_state.current_question['answer']}. Great job! Your score is now {self.trivia_state.score} out of {self.trivia_state.questions_asked}."
        else:
            feedback = f"Sorry, that's not quite right. The correct answer is {self.trivia_state.current_question['answer']}. Your score is {self.trivia_state.score} out of {self.trivia_state.questions_asked}."
        
        logger.info(f"User answered: {user_answer} - {'Correct' if is_correct else 'Incorrect'}")
        logger.info(f"Current score: {self.trivia_state.score}/{self.trivia_state.questions_asked}")
        
        # Update backend
        await self.update_backend_score()
        
        return None, feedback
    
    @function_tool
    async def get_current_score(self):
        """Get the user's current score"""
        return None, f"Your current score is {self.trivia_state.score} out of {self.trivia_state.questions_asked} questions."
    
    @function_tool
    async def end_game(self):
        """End the trivia game and show final score"""
        self.trivia_state.game_active = False
        percentage = (self.trivia_state.score / self.trivia_state.questions_asked * 100) if self.trivia_state.questions_asked > 0 else 0
        
        final_message = f"Thanks for playing! Your final score is {self.trivia_state.score} out of {self.trivia_state.questions_asked}. That's {percentage:.1f} percent correct. Great job!"
        
        await self.update_backend_score()
        
        return None, final_message
    
    async def on_enter(self):
        """Called when agent enters the room"""
        await self.session.say(
            """Welcome to Trivia Time! I'm your host and I'm excited to test your knowledge. 
            I'll ask you questions from different categories like Science, History, Geography, Movies, and Sports. 
            Just answer with what you think is correct. Ready to start? Let me ask you the first question!""",
            allow_interruptions=False
        )
        
        # Ask first question automatically
        await asyncio.sleep(0.5)
        question_data = self.trivia_state.get_next_question()
        question_text = f"Question number {self.trivia_state.questions_asked}. Category: {question_data['category']}. {question_data['question']}"
        
        logger.info(f"First question: {question_data['question']}")
        logger.info(f"Correct answer: {question_data['answer']}")
        
        await self.session.say(question_text)


# Create agent server
server = AgentServer()

@server.rtc_session()
async def trivia_session(ctx: agents.JobContext):
    """Main entry point for the trivia agent"""
    
    # Create trivia state for this session
    trivia_state = TriviaState()
    
    # Get participant identity if available
    await ctx.connect()
    if ctx.room.remote_participants:
        participant = list(ctx.room.remote_participants.values())[0]
        trivia_state.participant_identity = participant.identity
        logger.info(f"Participant joined: {participant.identity}")
    
    # Create agent session with OpenAI Realtime
    session = AgentSession(
        llm=openai.realtime.RealtimeModel(
            voice="alloy",  # Friendly voice for trivia host
            temperature=0.8,
            instructions="You are an enthusiastic trivia game host. Keep responses natural and conversational."
        )
    )
    
    # Start the session
    await session.start(
        room=ctx.room,
        agent=TriviaHostAgent(trivia_state),
        room_options=room_io.RoomOptions(
            audio_input=room_io.AudioInputOptions(
                noise_cancellation=noise_cancellation.BVC(),
            ),
        ),
    )
    
    logger.info("Trivia agent session started")


if __name__ == "__main__":
    agents.cli.run_app(server)

